app/controllers/api/base_controller/renderer.rb

```diff
@@ -125,9 +125,10 @@ def virtual_attribute_accessor(type, attr)
 
       private
 
-      def resource_search(id, type, klass)
-        validate_id(id, type, klass)
-        key_id = collection_config.resource_identifier(type)
+      def resource_search(id, type, klass = nil, key_id = nil)
+        klass  ||= collection_class(type)
+        key_id ||= collection_config.resource_identifier(type)
```

Comment:
Instead of making key_id a parameter, which is derived in every case except 1, perhaps we should make it a subclassable method, similar to validate_id, which is also only overridden in one case.  In fact, maybe it makes sense to move the "key_id" concept into validate_id, perhaps having that method _return_ the `key_id`

Reply:
Yea, I think we only have 2 or 3 cases where we pass in a string or an id.

I liked passing in id here because the method that calls this has the knowledge that a string name is being used to look up. So it consolidates the logic.

I like the idea of deferring this decision.

Reply:
The over arching goal is to set this up in a way that `api_resource` can just call `resource_search` without any special cases and it would work.

Currently, `resource_search` for these cases require `data['name']`. This is not available so that is why there is all this kludging.

Maybe once klass (and possibly key_id) is gone, then that will open us up for passing in the `data` or `name` parameter and they will all fall into the common pattern.